File: ocsgw/src/main/java/org/ostelco/ocsgw/utils/AppConfig.java
package org.ostelco.ocsgw.utils;

import org.ostelco.ocsgw.datasource.DataSourceType;
import org.ostelco.ocsgw.datasource.SecondaryDataSourceType;

import java.io.IOException;
import java.io.InputStream;
import java.util.Properties;

public class AppConfig {

    private final Properties prop = new Properties();

    public AppConfig() throws IOException {
        final String fileName = "config.properties";
        InputStream iStream = this.getClass().getClassLoader().getResourceAsStream(fileName);
        prop.load(iStream);
        iStream.close();
    }

    public DataSourceType getDataStoreType () {
        // OCS_DATASOURCE_TYPE env has higher preference over config.properties
        final String dataSource = System.getenv("OCS_DATASOURCE_TYPE");
        if (dataSource == null || dataSource.isEmpty()) {
            try {
                return DataSourceType.valueOf(prop.getProperty("DataStoreType", "Local"));
            } catch (IllegalArgumentException e) {
                return DataSourceType.Local;
            }
        }
        try {
            return DataSourceType.valueOf(dataSource);
        } catch (IllegalArgumentException e) {
            return DataSourceType.Local;
        }
    }

    public Long getDefaultRequestedServiceUnit () {
          final String defaultRequestedServiceUnit = System.getProperty("DEFAULT_REQUESTED_SERVICE_UNIT");
          if (defaultRequestedServiceUnit == null || defaultRequestedServiceUnit.isEmpty()) {
              return 40_000_000L;
          } else {
              return Long.parseLong(defaultRequestedServiceUnit);
          }
    }

    public SecondaryDataSourceType getSecondaryDataStoreType () {
        // OCS_SECONDARY_DATASOURCE_TYPE env has higher preference over config.properties
        final String secondaryDataSource = System.getenv("OCS_SECONDARY_DATASOURCE_TYPE");
        if (secondaryDataSource == null || secondaryDataSource.isEmpty()) {
            try {
                return SecondaryDataSourceType.valueOf(prop.getProperty("SecondaryDataStoreType", "PubSub"));
            } catch (IllegalArgumentException e) {
                return SecondaryDataSourceType.PubSub;
            }
        }
        try {
            return SecondaryDataSourceType.valueOf(secondaryDataSource);
        } catch (IllegalArgumentException e) {
            return SecondaryDataSourceType.PubSub;
        }
    }

    public String getGrpcServer() {
        return getEnvProperty("OCS_GRPC_SERVER");
    }

    public String getPubSubProjectId() {
        return getEnvProperty("PUBSUB_PROJECT_ID");
    }

    public String getPubSubTopicIdForCcr() {
        return getEnvProperty("PUBSUB_CCR_TOPIC_ID");
    }

    public String getPubSubTopicIdForCca() {
        return getEnvProperty("PUBSUB_CCA_TOPIC_ID");
    }

    public String getPubSubSubscriptionIdForCca() {
        return getEnvProperty("PUBSUB_CCA_SUBSCRIPTION_ID");
    }

    public String getPubSubSubscriptionIdForActivate() {
        return getEnvProperty("PUBSUB_ACTIVATE_SUBSCRIPTION_ID");
    }


    private String getEnvProperty(String propertyName) {
        final String value = System.getenv(propertyName);
        if (value == null || value.isEmpty()) {
            throw new Error("No "+ propertyName + " set in env");
        }
        return value;
    }
}

File: ocsgw/src/main/kotlin/org/ostelco/ocsgw/utils/AppConfig.kt
package org.ostelco.ocsgw.utils

import org.ostelco.diameter.getLogger
import org.ostelco.ocsgw.datasource.DataSourceType
import org.ostelco.ocsgw.datasource.SecondaryDataSourceType
import java.util.*

class AppConfig {

    private val logger by getLogger()

    private val prop = Properties()
    // OCS_DATASOURCE_TYPE env has higher preference over config.properties
    val dataStoreType: DataSourceType
        get() { // OCS_DATASOURCE_TYPE env has higher preference over config.properties
            val dataSource = System.getenv("OCS_DATASOURCE_TYPE")
            return if (dataSource.isNullOrEmpty()) {
                try {
                    DataSourceType.valueOf(prop.getProperty("DataStoreType", "Local"))
                } catch (e: IllegalArgumentException) {
                    logger.warn("Could not get the DataSourceType from property")
                    DataSourceType.Local
                }
            } else try {
                DataSourceType.valueOf(dataSource)
            } catch (e: IllegalArgumentException) {
                logger.warn("Could not get the DataSourceType from env")
                DataSourceType.Local
            }
        }

    val defaultRequestedServiceUnit: Long
        get() {
            val defaultRequestedServiceUnit = System.getProperty("DEFAULT_REQUESTED_SERVICE_UNIT")
            return if (defaultRequestedServiceUnit.isNullOrEmpty()) {
                40000000L
            } else {
                defaultRequestedServiceUnit.toLong()
            }
        }

    // OCS_SECONDARY_DATASOURCE_TYPE env has higher preference over config.properties
    val secondaryDataStoreType: SecondaryDataSourceType
        get() { // OCS_SECONDARY_DATASOURCE_TYPE env has higher preference over config.properties
            val secondaryDataSource = System.getenv("OCS_SECONDARY_DATASOURCE_TYPE")
            return if (secondaryDataSource.isNullOrEmpty()) {
                try {
                    SecondaryDataSourceType.valueOf(prop.getProperty("SecondaryDataStoreType", "PubSub"))
                } catch (e: IllegalArgumentException) {
                    SecondaryDataSourceType.PubSub
                }
            } else try {
                SecondaryDataSourceType.valueOf(secondaryDataSource)
            } catch (e: IllegalArgumentException) {
                logger.warn("Could not get the Secondary DataSourceType")
                SecondaryDataSourceType.PubSub
            }
        }

    val grpcServer: String
        get() = getEnvProperty("OCS_GRPC_SERVER")

    val pubSubProjectId: String
        get() = getEnvProperty("PUBSUB_PROJECT_ID")

    val pubSubTopicIdForCcr: String
        get() = getEnvProperty("PUBSUB_CCR_TOPIC_ID")

    val pubSubTopicIdForCca: String
        get() = getEnvProperty("PUBSUB_CCA_TOPIC_ID")

    val pubSubSubscriptionIdForCca: String
        get() = getEnvProperty("PUBSUB_CCA_SUBSCRIPTION_ID")

    val pubSubSubscriptionIdForActivate: String
        get() = getEnvProperty("PUBSUB_ACTIVATE_SUBSCRIPTION_ID")

    private fun getEnvProperty(propertyName: String): String {
        val value = System.getenv(propertyName)
        if (value.isNullOrEmpty()) {
            throw Error("No $propertyName set in env")
        }
        return value
    }

    init {
        val fileName = "config.properties"
        val iStream = this.javaClass.classLoader.getResourceAsStream(fileName)
        prop.load(iStream)
        iStream?.close()
    }
}